Make --library-python defaults more Python friendly (#15475)
chapel-lang/chapel: doc/rst/technotes/libraries.rst

   :ref:`readme-chplenv.CHPL_COMPILER` for more information on the values of
   ``CHPL_TARGET_COMPILER``

.. _Python_Output_Directory_Name:

Python Output Directory Name
----------------------------

By default, the name of the directory created to contain the generated Python
module will match the generated Python module name. To change the
output directory name so that it does not match the generated Python module
name, use the compilation flag ``--library-dir``.

.. code-block:: bash

  # Builds Python module as foo/foo.py from foo.chpl

>   chpl --library --library-python foo.chpl

Reviewer: Is `--library` necessary in these examples?

```suggestion
  chpl --library --library-python foo.chpl
```

>    export PYTHONPATH=lib/:$PYTHONPATH

Reviewer 2: Is this step necessary by default now, or does the `__init__.py` file change that?

Reviewer 2: The tests seem to explicitly unset the PYTHONPATH, so it seems like we should indicate it isn't necessary?

Author: If you create an init file then you can just import the entire directory as a package, which works like importing any other Python package. I left the old paragraphs because it's still a valid way of doing things if one doesn't want to make use of the init file. Should I just remove it?

Reviewer 2: I think keeping it is the right call, since there are some strategies that still require it, but I'd be more explicit about when those situations occur

prior to any Chapel library function calls.

Also unlike the C case, the clean up function is called ``chpl_cleanup()``.
This function will still need to be called after all the Chapel library
function calls are finished, unless you have imported the output directory

> as a package.

Reviewer: Maybe a link to the next section would make this clearer:

```suggestion
as a package, using the :ref:`Python Init File`.
```

Reviewer: (not sure if that's the right syntax)

   Make sure your Python functionality is also complete before calling this
   function.

.. note::

  If you are taking advantage of the generated ``__init__.py`` initializer
  file to import the output directory as a package, you do not need to call
  ``chpl_cleanup()`` yourself because it is already registered to be called
  at program exit. The generated initializer is explained below.

Python Init File
----------------

A simple ``__init__.py`` file is generated in the output directory along with

> the Python module. It looks roughly like the following:

Reviewer: Good call on the use of roughly, given that this can change over time.

Author: Thanks!

.. code-block:: python

  import atexit

  #
  # Here directoryName is the name of the directory containing your Python
  # module, and moduleName is the name of your Python module.
  #
  from directoryName.moduleName import *

  atexit.register(moduleName.chpl_cleanup)

> The initializer file makes working with generated Python code easier than

Reviewer: This sentence seems unnecessary, but it does make me want to use this feature! :D

ever by transforming the output directory into a Python package. The
initializer file will also register ``chpl_cleanup()`` to be called
automatically at program exit.

> To take advantage of these new features, the directory containing the

Reviewer: Can we make this paragraph more concise? I would also drop "new" to make this documentation a little more timeless.

Reviewer: As a starting point:

> To utilize the generated `__init__.py`, import the generated directory (`foo/`) as a python package rather than the python module (`foo/foo.py`), as demonstrated below.

Author: Mind reviewing my edit?

generated Python module needs to be visible to the script that imports it
in some fashion, whether that be by making it a child package or by
placing the package directory on your ``PYTHONPATH`` environment variable.

.. code-block:: bash

  # Builds Python module as foo/foo.py
  chpl --library-python foo.chpl

  # Adds the current directory to your PYTHONPATH

>   export PYTHONPATH="$PWD"

Reviewer: Users may want to prepend to their python path more commonly:

```suggestion
  export PYTHONPATH="$PWD:$PYTHONPATH"
```

Author: Can't believe I forgot to do that. Thanks!

   Make sure your Python functionality is also complete before calling this
   function.

.. note::

  If you are taking advantage of the generated ``__init__.py`` initializer
  file to import the output directory as a package, you do not need to call
  ``chpl_cleanup()`` yourself because it is already registered to be called
  at program exit. The generated initializer is explained below.

.. _Python_Init_File:

Python Init File
----------------

A simple ``__init__.py`` file is generated in the output directory along with
the Python module. It looks roughly like the following:

.. code-block:: python

  import atexit

  #
  # Here directoryName is the name of the directory containing your Python
  # module, and moduleName is the name of your Python module.
  #
  from directoryName.moduleName import *

  atexit.register(moduleName.chpl_cleanup)

> The initializer file transforms the output directory into a Python package.

Reviewer: ```suggestion
The initializer file allows the output directory to be imported as a Python package.
```

It will also register ``chpl_cleanup()`` to be called automatically at
program exit.

> The directory containing generated Python code must be visible to any script

Reviewer: I think our attempt to summarize the python import system makes this paragraph harder to follow than is has to be. Maybe we could just list the most common approaches (importing locally or through `PYTHONPATH`) and then link the python docs for more details, e.g.

```
Like any other Python package, the generated Python package must be visible 
in order to import it, such as through importing it locally or adding the path to ``PYTHONPATH``.
Refer to the `Python import documentation`_ for more details on python import visibility.

.. _Python import documentation https://docs.python.org/3/reference/import.html#the-import-system
```

that attempts to import it as a package. One option is to place the directory
somewhere in an existing codebase and use an absolute import. Another option
is to place the package directory on your ``PYTHONPATH`` environment variable.

.. code-block:: bash

  # Builds Python module as foo/foo.py
  chpl --library-python foo.chpl

  # Adds the current directory to your PYTHONPATH
  export PYTHONPATH="$PWD:$PYTHONPATH"

From within your Python script:

.. code-block:: Python

  import foo

  #
  # Setup and use foo as normal. Note that we no longer have to call
  # ``chpl_cleanup()`` when we are finished.
  #
  foo.chpl_setup()
  foo.baz(2)

.. note::

  The Chapel compiler will not generate an initializer file if a file with
  the name ``__init__.py`` already exists in the output directory. The

>   compiler will emit a warning instead.

Reviewer: Is there a way to disable this warning for projects that intentionally utilize this behavior?

Author: Sounds like maybe I _should_ add a compiler flag?

Reviewer: I would hold off on creating a one-off compiler flag for this warning control. There are more organized efforts that could enable this in the future:

- different categories of warnings (https://github.com/chapel-lang/chapel/issues/11576)
- generalized warning control (https://github.com/chapel-lang/chapel/issues/11576)
